mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,175 @@
+fn send_command(port: *mut c_void, command: Command) -> Result<(), String> {
+    let bytes = command.as_bytes();
+    unsafe {
+        let written = sp_blocking_write(port, bytes.as_ptr(), bytes.len(), 1000);
+        if written != bytes.len() as i32 {
+            return Err(format!(
+                "{:?} failed to send {:?}: {}",
+                port, command, written
+            ));
+        }
+    }
+    Ok(())
+}
+
+fn read_controller_state(port: *mut c_void) -> Result<ControllerState, String> {
+    let mut buffer = [0u8; 3];
+    unsafe {
+        let bytes_read = sp_blocking_read(port, buffer.as_mut_ptr(), buffer.len(), 1000);
+        if bytes_read != buffer.len() as i32 {
+            return Err(format!(
+                "{:?} failed to read controller state: {}",
+                port, bytes_read
+            ));
+        }
+    }
+    let adc_value = u16::from_be_bytes([buffer[1], buffer[2]]);
+    let buttons = buffer[0];
+    // println!("Read buttons: {:08b}, ADC value: {}", buttons, adc_value); // Debug print
+    Ok(ControllerState {
+        buttons,
+        adc_value,
+    })
+}
+
+pub fn controller(
+    port_name: &str,
+    receiver: Receiver<(String, String)>,
+    model_sender: Sender<(String, String)>,
+    player: usize,
+) {
+    let port = match open_port(port_name) {
+        Ok(port) => port,
+        Err(e) => {
+            eprintln!("{}", e);
+            return;
+        }
+    };
+    model_sender
+        .send((player.to_string(), "".to_string()))
+        .unwrap();
+
+    let mut last_button_state = 0;
+    let mut last_debounce_time = Instant::now();
+    let debounce_delay = std::time::Duration::from_millis(50);
+
+    loop {
+        // Wait for game to start
+        if let Ok((_, message)) = receiver.recv() {
+            if message == "Escape" {
+                break;
+            }
+        }
+
+        // Start LED command sequence
+        loop {
+            if let Ok((_, message)) = receiver.recv() {
+                if let Err(e) = send_command(port, Command::from_string(&message).unwrap()) {
+                    eprintln!("Failed to send command: {}", e);
+                };
+                if message == "Start" {
+                    break;
+                }
+            }
+        }
+
+        // Read controller state
+        loop {
+            match receiver.try_recv() {
+                Ok((_, message)) => {
+                    if message == "Game over" {
+                        break;
+                    }
+                }
+                Err(_) => {}
+            }
+            let controller_state = match read_controller_state(port) {
+                Ok(state) => state,
+                Err(e) => {
+                    eprintln!("{}", e);
+                    break;
+                }
+            };
+
+            // Debounce button state
+            if controller_state.buttons != last_button_state {
+                last_debounce_time = Instant::now();
+            }
```

Comment:
nit: would be more correct to debounce individual buttons, but this is almost the same

Reply:
oh actually, because of how you've implemented debounce, the button presses never get registered